### PR TITLE
Spring Boot Portfolio Project — Part 7. JWT. Executor: Oleksii Bashkirov.

### PR DESCRIPTION
**Pull request by OleksiiBashkirov**
feat: Implement JWT authentication

- Added `login(UserLoginRequestDto request)` method to `AuthenticationController`.
- Created `AuthenticationService`, `JwtAuthenticationFilter`, and `JwtUtil` classes.
- Made `/auth/login` endpoint publicly accessible.
- Configured `SecurityFilterChain` to use the new JWT filter.
- Implemented data validation for new DTO classes.

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -101,6 +102,31 @@
...
+            <version>${jjwt.version}</version>
+            <scope>runtime</scope>
+        </dependency>
+        <dependency>
+            <groupId>io.jsonwebtoken</groupId>
+            <artifactId>jjwt-jackson</artifactId>
+            <version>${jjwt.version}</version>
+            <scope>runtime</scope>
+        </dependency>
+
```
Comment (Elena-Bruyako) [inline pom.xml]: remove empty lines

File: src/main/java/mate/academy/config/JwtConfig.java [hunk 2]
```diff
@@ -0,0 +1,20 @@
+package mate.academy.config;
+
+import javax.crypto.SecretKey;
+import javax.crypto.spec.SecretKeySpec;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+
+@Configuration
+public class JwtConfig {
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/config/JwtConfig.java]: Why do you need this?

File: src/main/java/mate/academy/config/SecurityConfig.java [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+public class SecurityConfig {
+    private final CustomUserDetailsService customUserDetailsService;
+    private final JwtAuthenticationFilter jwtAuthenticationFilter;
+
+    @Bean
+    public PasswordEncoder passwordEncoder() {
+        return new BCryptPasswordEncoder();
+    }
+
+    @Bean
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/config/SecurityConfig.java]: This confi could be much simplier
```suggestion
    @Bean
    public SecurityFilterChain securityFilterChain(HttpSecurity http) throws Exception {
        return http
                .cors(AbstractHttpConfigurer::disable)
                .csrf(AbstractHttpConfigurer::disable)
                .authorizeHttpRequests(
                        auth -> auth
                                .requestMatchers(
                                        antMatcher("/auth/**"),
                                        antMatcher("/swagger-ui/**"),
                                        antMatcher("/v3/api-docs/**"))
                                .permitAll()
                                .anyRequest()
                                .authenticated()
                )
                .sessionManagement(
                        s -> s.sessionCreationPolicy(SessionCreationPolicy.STATELESS))
                .addFilterBefore(jwtAuthFilter,
                        UsernamePasswordAuthenticationFilter.class)
                .userDetailsService(service)
                .build();
    }

    @Bean
    public AuthenticationManager authenticationManager(
            AuthenticationConfiguration authenticationConfiguration) throws Exception {
        return authenticationConfiguration.getAuthenticationManager();
    }
```

File: src/main/java/mate/academy/config/SwaggerConfig.java [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+
+@Configuration
+public class SwaggerConfig {
+
+    @Bean
+    public OpenAPI customOpenApi() {
+        return new OpenAPI()
+                .components(new Components().addSecuritySchemes("BearerAuth",
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/config/SwaggerConfig.java]: create constant for "BearerAuth"

File: src/main/java/mate/academy/dto/UserLoginResponseDto.java [hunk 5]
```diff
@@ -0,0 +1,10 @@
+package mate.academy.dto;
+
+import lombok.AllArgsConstructor;
+import lombok.Data;
+
+@Data
+@AllArgsConstructor
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/dto/UserLoginResponseDto.java]:
```suggestion
```

File: src/main/java/mate/academy/repository/UserRepository.java [hunk 6]
```diff
@@ -1,10 +1,11 @@
 package mate.academy.repository;
 
+import java.util.Optional;
 import mate.academy.model.User;
 import org.springframework.data.jpa.repository.JpaRepository;
-import org.springframework.stereotype.Repository;
 
-@Repository
 public interface UserRepository extends JpaRepository<User, Long> {
+    Optional<User> findByEmail(String email);
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/repository/UserRepository.java:10]:
```suggestion
    @EntityGraph(attributePaths = "roles")
    Optional<User> findByEmail(String email);
```

File: src/main/java/mate/academy/security/JwtAuthenticationFilter.java [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+
+        if (isTokenValid(token)) {
+            setAuthenticationContext(token);
+        }
+
+        filterChain.doFilter(request, response);
+    }
+
+    private String getToken(HttpServletRequest httpServletRequest) {
+        String bearerToken = httpServletRequest.getHeader("Authorization");
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/security/JwtAuthenticationFilter.java]:
```suggestion
        String bearerToken = httpServletRequest.getHeader(HttpHeaders.AUTHORIZATION);
```

File: src/main/java/mate/academy/security/JwtAuthenticationFilter.java [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+        if (StringUtils.hasText(bearerToken) && bearerToken.startsWith("Bearer ")) {
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/security/JwtAuthenticationFilter.java]: Create constant for "Bearer "

File: src/main/java/mate/academy/security/JwtAuthenticationFilter.java [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+            return bearerToken.substring(7);
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/security/JwtAuthenticationFilter.java]:
```suggestion
            return bearerToken.substring(constant.length());
```

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see comments

**Comment by OleksiiBashkirov** [comment]
I wanted to apologize for my recent accidental pull request to https://github.com/OleksiiBashkirov/online-book-store/pull/8. 
Due to an accidental click, I submitted the changes meant for the "JWT" theme under the "Spring Boot Security" theme.
Sorry about that. Please disregerd that pull request.
Thank you for your understanding and patience!

File: src/main/java/mate/academy/config/SwaggerConfig.java [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+    private static final String BEARER_AUTH = "BearerAuth";
+
+    @Bean
+    public OpenAPI customOpenApi() {
+        return new OpenAPI()
+                .components(new Components().addSecuritySchemes(BEARER_AUTH,
+                        new SecurityScheme()
+                                .type(SecurityScheme.Type.HTTP)
+                                .scheme("bearer")
+                                .bearerFormat("JWT")))
```
Comment (MentorMateAcademy) [inline src/main/java/mate/academy/config/SwaggerConfig.java]: you can save to constant too

**Review by MentorMateAcademy — CHANGES_REQUESTED** [review]

**Review by MentorMateAcademy — APPROVED** [review]

File: src/main/java/mate/academy/controller/AuthenticationController.java [hunk 11]
```diff
@@ -16,10 +19,16 @@
...
     private final UserService userService;
+    private final AuthenticationService authenticationService;
 
     @PostMapping("/registration")
     public UserResponseDto register(@RequestBody @Valid UserRegistrationRequestDto requestDto)
             throws RegistrationException {
         return userService.register(requestDto);
     }
+
+    @PostMapping("/login")
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/controller/AuthenticationController.java:30]: Add swagger annotations

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
This branch has conflicts, please fix them

**Pull request merged**